<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>

Привет Александр! Меня зовут Марат, и я буду твоим ревьюером. Спешу сообщить что все ключевые этапы в работе выполнены,  с задачей тебе удалось справиться. По поводу обращения - в IT сфере принято общаться на «ты» :) Но, если привычней на «вы», дай знать. Как ревьюера моя задача помочь тебе в развитии, дав хорошие советы. Я внимательно посмотрю твой код, ознакомлюсь с твоими выводами и оставлю комментарии. Где то могу предложить небольшие исправление в коде, но ненавязчиво. Где потребуются уточнения, я оставлю много наводящих вопросов. Они помогут тебя с поиском верного решения.

Все мои комментарии размечены по цветам, для лучшего восприятия. 
    
<div class="alert alert-success">Зеленым цветом и словом «Успех» отмечены особо удачные и элегантные решения, которыми ты можешь гордиться. </div>
        
<div class="alert alert-warning">Желтым и значком словом «Совет», помечены решения у которых есть альтернативные решения, более оптимальные. Ты можешь найти их сразу и доработать проект, или отложить это на потом, для будущих проектах. Проект будет принят и без их доработки. </div>
        
<div class="alert alert-danger"> Красным цветом и значком словом «Ошибка» помечу твои решения, на которые стоит обратить внимание прежде всего. После их доработки проект будет принят. </div>
        
Залог успеха - работа сообща, взаимное уважение и работа в диалоге. Поэтому, помечай свои ответные комментарии на мои реплики заметным цветом или курсивом, так мне будет легче их отслеживать. Пожалуйста, не изменяй и не удаляй мои комментарии. Все это поможет выполнить повторную проверку быстрей.

</div>

# Анализ развития игровой индустрии: платформы, жанры и предпочтения игроков

- Автор: Максимов А.В
- Дата: 04.12.2024

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>



Совет 🤔:


Вот тут нас прямо просят написать заголовок проекта.  Некоторые Reviewer   за это даже ставят красное )

### Цели и задачи проекта

**Цель проекта:**
- Изучить развитие игровой индустрии


**Задачи:**
- Познакомиться с данными
- Проверить их корректность
- Провести предобработку,получив необходимый срез данных.

### Описание данных

**`Name`** - название игры


**`Platform`** - название платформы


**`Year of Release`** - год выпуска игры


**`Genre`** - жанр игры


**`NA sales`** - продажи в Северной Америке (в миллионах проданных копий)


**`EU sales`** - продажи в Европе (в миллионах проданных копий)


**`JP sales`** - продажи в Японии (в миллионах проданных копий)


**`Other sales`** - продажи в других странах (в миллионах проданных копий).


**`Critic Score`** - продажи в других странах (в миллионах проданных копий).


**`User Score`** - оценка пользователей (от 0 до 10).


**`Rating`** - рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

***Дополнительная информация о столбце `Rating`:***


- **RP (Rating Pending)** — используется в рекламных материалах для игр, которым ещё не присвоен окончательный рейтинг ESRB. Принят в 1994 году и действует до сих пор.


- **EC (Early Childhood)** — игры, ориентированные на дошкольную аудиторию. Не содержат материалов, которые родители могли бы считать неприемлемыми. Принят в 1994 году и действовал до 2018 года.


- **E (Everyone)** — игры для всех возрастов. Могут содержать незначительное количество насилия и сленга. До 1998 года рейтинг был известен как Kids to Adults (K-A). Принят в 1994 году и действует до сих пор.


- **E10+ (Everyone 10 and older)** — игры для людей от 10 лет и старше. Могут содержать более высокий уровень насилия, грубую разговорную речь. Принят в 2005 году и действует до сих пор.


- **T (Teen)** — игры для людей от 13 лет и старше. Могут содержать умеренное количество насилия, включая небольшие количества крови, грубую разговорную речь и юмор, сексуальное содержание. Принят в 1994 году и действует до сих пор.


- **M (Mature)** — игры для лиц от 17 лет и старше. Могут содержать более интенсивные или реалистичные изображения насилия, включая кровь, ужасы, увечья и изображения смерти, непристойные сексуальные сцены, частичную наготу и грубую ненормативную лексику. Принят в 1994 году и действует до сих пор.


- **AO (Adults Only)** — игры для людей от 18 лет и старше. Содержат материалы с более высоким уровнем воздействия, чем рейтинг M. Принят в 1994 году и действует до сих пор.

## Содержимое проекта

**1.Загрузка и знакомство с данными**
- Импорт библиотек
- Загрузка данных
- Вывод первых строк данных
- Получение информации о датасете


**2.Анализ данных**
- Проверка названий столбцов
- Приведение названий к стандартному виду
- Проверка типов данных


**3.Обработка пропусков**
- Подсчет пропусков
- Анализ данных с пропусками
- Обработка пропусков


**4.Устранение дубликатов**
- Проверка уникальных значений
- Нормализация текстов
- Удаление явных дубликатов
- Отчет об удаленных строках

**5.Фильтрация данных**
- Фильтрация по годам


**6.Категоризация данных**
- Категоризация оценок пользователей
- Категоризация оценок критиков


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



Отличное введение в проект,  сразу дал расшифровки для     данных что   пригодятся в проекте


 

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
import pandas as pd
import numpy as np
# Пустые ячейки после каждого задания — примерное пространство для работы.
# Вы можете свободно добавлять или удалять ячейки по своему усмотрению в зависимости от логики и объёма работы.


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Собираем все импорты в верхней части, чтобы легче было ориентироваться и добавлять новые по необходимости. Пока у нас пока у нас она одна, но в будущем будет много) 


<div class="alert alert-warning">


Совет:

   
 

- всё-таки стоит импорт библиотек и выгрузку данных разнести в разные ячейки, так будет стилистически правильней  


    
    
- есть рекомендации PEP-8 при написании кода, в том числе и для импортов. Если интересно можешь почитать [тут](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html). Будущим пригодится

  

In [2]:
df=pd.read_csv('https://code.s3.yandex.net//datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
display(df.head())
df.info()

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>



Ошибка ❌:



df.info() это классика при предварительном обзоре данных, но не забываем вывести тут же на экран наши данные.  Обычно это делают используя head, но для разнообразия можешь попробовать sample. Всё-таки стоит получить предварительное представление о данных и таким образом - посмотреть что вообще представляют наши данные. Это Например Мы точно не знаем Rating это числа, или текст.  Если числа то object ошибка, если буквы, то данные прочитали верно

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



Исправлено


 

Колонки EU и JP sales надо привести к float, так как продажи измеряются в цифрах, и они последуюатся в будущем для подсчета общих продаж
Колонку User Score требуется привести к типу данных float, так как пользователидают оценку игре в целочисленном формате от 0 до 10 зачастую

В нашем датафрейме 16956 значений, и судя по нашим столбцам, в некоторых из них есть пустые значения. Например в колонке Name, там пропущно 2 значения,
а у столбца Year of Release пропущено jколо 300 значений, с пропусками нужно поработать

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



Верные наблюдения


 

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

<font color='#777778'>Используйте ячейки типа Markdown для промежуточных выводов и расширенных комментариев к действиям с данными. </font>

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [4]:
#Делаем нижний регистр для всех столбцов и заменяем пробел на нижнее подчекривание
df.columns = df.columns.str.lower().str.replace(' ','_')
display(df.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')




<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

👍 помнишь о стилистике. Молодец


И здорово что используешь автоматизированный подход, причём выбрав оптимальный вариант. Всё-таки впереди нас могут ждать проекты со многими десятками или сотнями столбцов, и было бы затруднительно переименовать их руками набивая названия. Молодец
    
    
 <div class="alert alert-warning">


Совет: 


Есть и другие способы автоматизированного приведения к змеиному виду, например [регуляризация]https://habr.com/ru/articles/349860/.   Она пригодится если .lower().str.replace окажется недостаточным. Например если названия столбцов будут в верблюжьем (CamelCase) стиле, с помощью регуляризации легко перевести из верблюжьего в змеиный  
     
     
     df.columns = [re.sub(r'(?<!^)(?=[A-Z])', '_', i).lower() for i in df.columns]
     
     
     

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.
- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>



Ошибка ❌:

 
    
Действия ниже верные, но предварительно  предлагаю тебе провести   анализ и выяснить, почему числовые данные имеют не числовой тип данных.  Для этого достаточно вывести уникальные значения по   столбцам где мы видим несоответствия, и проанализировать увиденное





In [5]:
# Проверяем уникальные значения в проблемных столбцах
for column in ['eu_sales', 'jp_sales', 'year_of_release', 'user_score','critic_score']:
    unique_values = df[column].unique()
    print(f"Уникальные значения в столбце {column}: {unique_values}")

Уникальные значения в столбце eu_sales: ['28.96' '3.58' '12.76' '10.93' '8.89' '2.26' '9.14' '9.18' '6.94' '0.63'
 '10.95' '7.47' '6.18' '8.03' '4.89' '8.49' '9.09' '0.4' '3.75' '9.2'
 '4.46' '2.71' '3.44' '5.14' '5.49' '3.9' '5.35' '3.17' '5.09' '4.24'
 '5.04' '5.86' '3.68' '4.19' '5.73' '3.59' '4.51' '2.55' '4.02' '4.37'
 '6.31' '3.45' '2.81' '2.85' '3.49' '0.01' '3.35' '2.04' '3.07' '3.87'
 '3.0' '4.82' '3.64' '2.15' '3.69' '2.65' '2.56' '3.11' '3.14' '1.94'
 '1.95' '2.47' '2.28' '3.42' '3.63' '2.36' '1.71' '1.85' '2.79' '1.24'
 '6.12' '1.53' '3.47' '2.24' '5.01' '2.01' '1.72' '2.07' '6.42' '3.86'
 '0.45' '3.48' '1.89' '5.75' '2.17' '1.37' '2.35' '1.18' '2.11' '1.88'
 '2.83' '2.99' '2.89' '3.27' '2.22' '2.14' '1.45' '1.75' '1.04' '1.77'
 '3.02' '2.75' '2.16' '1.9' '2.59' '2.2' '4.3' '0.93' '2.53' '2.52' '1.79'
 '1.3' '2.6' '1.58' '1.2' '1.56' '1.34' '1.26' '0.83' '6.21' '2.8' '1.59'
 '1.73' '4.33' '1.83' '0.0' '2.18' '1.98' '1.47' '0.67' '1.55' '1.91'
 '0.69' '0.6' '1.93' '1.64' '0.

Можем заметить, что в столбцах по продаже **jp_sales и eu_sales** есть значения не цифрового формата *'unknown'*, остальные значения - это цифры, но сами стобцы имеют типа данных *'object'*, что нам для анализа не подходит, стоит перевести эти слобцы в тип float, с заполнением не числовых значений как *'nan'*. В оставшихся трех столбцах, которые мы проверяли, **year_of_release, user_score и critic_score** можем заметить, что в первом случае мы видим пропуски все так же в виде значения *'nan'*,во втором случае, оценки пользователей указаны в числовом формате, но тип даннах этого столбца является *'object'*. Так же переведем эти слобцы в тип float, с заполнением не числовых значений как *'-1'*,эти значения в будущем нам не помешают при расчетах и они будут уникальными, так как до замены ни одного значения *'-1'* мы не видим. В последней ситуации тип данных в столбце **critic_score** уже является float, что нам в целом подходит, но есть значения *'nan'*, которые не являются числовыми, поэтому как и во втором случае, число *'-1'* ни разу не встречается, поэтому и преобразуем проуски *'nan* в *'-1'*

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍


 

In [6]:
#Перебираем нужные нам столбцы для проверки на пропуски и нужного нам типа данных
for column in ['na_sales','eu_sales','jp_sales','year_of_release','user_score']:
    df[column] = pd.to_numeric(df[column], errors = 'coerce')
#Задаем столбцу rating значение, которое указано в ТЗ
df['rating'] = df['rating'].fillna('NR')
df['user_score'] = df['user_score'].fillna(-1)
df['critic_score'] = df['critic_score'].fillna(-1)
display(df.head())
df.info()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0     28.96      3.77         8.45          76.0         8.0      E  
1      3.58      6.81         0.77          -1.0        -1.0     NR  
2     12.76      3.79         3.29          82.0         8.3      E  
3     10.93      3.28         2.95          80.0         8.0      E  
4      8.89     10.22         1.00          -1.0        -1.0     NR

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     16956 non-null  float64
 9   user_score       16956 non-null  float64
 10  rating           16956 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>



Совет 🤔:

print не всегда нужен, вывод  последней строки произойдет  и так, а когда нужен, вместо print лучше использовать display, будет красивей


Кстати ведь ты прекрасно используешь display, и вывод последней строки без print/display,  значит всё это знаешь и понимаешь.  Видимо тут ты по каким-то своим причинам используешь print?

***В ответ на комментарий выше, я просто эту часть кода написал, и забыл поменять + сыграл фактор, что сам проект доделывал ночью, не обратил внимания***

Выше мы поменяли тип данных на числовой в таких столбцах как: **eu_sales и jp_sales**. Данные в этих столбцах приводились в виде цифр, но в типе object. Предпологаю, что пропуски могли быть вызваны недоступностью данных на момент их сбора.Так же задали значения *-1*, столбцам **critic_score и user_score** изначально было задано ограничение в виде диапозона, который указан в описании данных, поэтому выбираем число за диапозоном обоих столбцов, для того, чтобы было удобнее производить расчеты.Последний столбец **rating** задал как *'NR'*, это будет означать, что эту пустое значение 

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



Здорово что есть обоснование, решение оптиимальное с использованием .to_numeric(df[column], errors = 'coerce'),  и рабочее с использованием заглушки и 'NR' (Ведь действительно отсутствие информации это тоже информация)
молодец что использовал цикл, всегда стоит  стремиться  сделать код лаконичный и читабельный

<div class="alert alert-warning">



Совет 🤔:

    
- работа с пропусками у нас в следующем разделе, так что заполнение стоило опустить ниже    
    
    
-  правильно что в качестве заглушки выбрал -1, но   предварительно стоило посмотреть, а нет ли такой цифры в данных?! Ведь у нас заглушка не должна принимать значения которые есть в реальных данных
    
    
    
- есть небольшой недостаток у когда мы используем заглушки -  мы всегда должны держать в голове что это не настоящие цифры. . И в какой-то момент мы можем... Забыть об этом. Поэтому есть ещё решение оставить всё как есть, потому что многие методы и функции питона отлично работает со столбцами в которых присутствует пропуски - могут прекрасно считать средние итд итп

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.
- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.
- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.


In [7]:
#Считаем абсолютное значение
missing_abs = df.isna().sum()
#Считаем относительное значение
missin_perc = (missing_abs / len(df))*100
print(missing_abs)
print(missin_perc)

name                 2
platform             0
year_of_release    275
genre                2
na_sales             0
eu_sales             6
jp_sales             4
other_sales          0
critic_score         0
user_score           0
rating               0
dtype: int64
name               0.011795
platform           0.000000
year_of_release    1.621845
genre              0.011795
na_sales           0.000000
eu_sales           0.035386
jp_sales           0.023590
other_sales        0.000000
critic_score       0.000000
user_score         0.000000
rating             0.000000
dtype: float64


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>



Совет 🤔:


- Получается ты выше заполнил пропуски, после чего посмотрел их количество. Ну ведь это стоило сделать ДО.  И тогда мы бы обнаружили огромное количество пропусков в рейтингах и оценках. как было бы логично и последовательно, как считаешь?
    
    
- Кроме того в задании нас просят объяснить возможные причины пропусков.  Тут можно даже провести небольшой анализ,  ведь можно обосновать наличие большого количества пропусков в рейтингах непопулярностью игры, и подтвердить это предположение смотрев на объёмы продаж 
    
    
 

- Возможно проще было воспринимать информацию, если сделать сортировочку, а вот так можно   раскрасить (+ округлить до сотых)

        
     .style.background_gradient(cmap='coolwarm', axis=1).set_precision(2)    
    
    
Выглядела бы симпатичнее    

Минимальное кол-во пропусков у столбцов: **name,genre,eu_sales,jp_sales**. У каждых из этих столбцов значения не переваливают за 10.Такие значения будут несущественные для дальнейшего анализа, поэтому эти строки можно удалить. Так же в столбце **year_of_release** тоже имеются пропуски, причем в разы больше, по сравнению с прошлыми. Данные пропуски являются всего 1.6% от всего количества данных в столбце **year_of_release**, поэтому эти данные тоже можно удалить.

В столбце **year_of_release** имеются пропуски потому что игра либо еще не вышло, либо год выпуска неизвестен. **critic_score и user_score** имеют пропуски, потому-что то оценки критиков и пользователей могли быть недоступны. А при столбцах name и genre они могли эксперементировать с названием игры и с жанром, поэтому могли и не выбрать до конца, либо данные были недоступны.  

In [8]:
#Удаляем у столбцов лишние данные
df.dropna(subset = ['name','genre','year_of_release','eu_sales','jp_sales'], inplace=True)
print('Кол-во пропусков до замены\n')
print(missing_abs, '\n')
#Заполняем пропуски
print('Кол-во пропусков после замены \n')
print(df.isna().sum())

Кол-во пропусков до замены

name                 2
platform             0
year_of_release    275
genre                2
na_sales             0
eu_sales             6
jp_sales             4
other_sales          0
critic_score         0
user_score           0
rating               0
dtype: int64 

Кол-во пропусков после замены 

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           0
jp_sales           0
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



решение обоснованное, здорово что проконтролировал что пропусков не осталось


<div class="alert alert-warning">


Совет 🤔:


Был вариант заполнить пропуски в продажах в группировке по  platform и year_of_release
    
 
Тут можно было написать функцию заполнения, после чего использовать apply    
    
Но тоже же самое можно сделать проще, нагляднее (меньше кода), эффективней (использование векторных операций), с помощью  [transform](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.transform.html#pandas-dataframe-transform)
    
    df['jp_sales'].fillna(df.groupby(['platform', 'year_of_release'])['jp_sales'].transform('mean')
    
    
Как видишь в отличие от  apply, для него даже не gридётся писать функцию, потому что он умеет работать с   встроенныеми функциямикак mean, sum, max, min      
    
    
Если есть желание, можешь попрактиковаться в использовании этого инструмента, и такого способа заполнения пропусков    

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>



Совет 🤔:


Было бы неплохо,  после заполнения пропусков, привести year_of_release и critic_score к целочисленным значениям.  Годы у нас не могут быть float как и critic_score в виде балов
    
 
Один из плюсов - это экономия ресурсов, ведь int, занимает гораздо меньше памяти чем float 

Кстати на просторах интернета можно найти  функцию для такого рода работы в автоматическом режиме, он подойдёт когда у нас будет большой датасет, и встанет вопрос как бы сэкономить ресурсы 
 

    # Оптимизация числовых типов данных
    def optimize_memory_usage(df: pd.DataFrame, print_size: bool=True) -> pd.DataFrame:
    
    
        '''
        Function optimizes memory usage in dataframe
        df: pd.DataFrame - data table
        print_size: bool - display of optimization results
        return pd.DataFrame - amount of optimized memory
        '''
    
    
        numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64'] # Типы, которые будем проверять на оптимизацию
        # Размер занимаемой памяти до оптимизации (в Мб)
        before_size = df.memory_usage().sum() / 1024**2    
        for column in df.columns:
            column_type = df[column].dtypes
            if column_type in numerics:
                column_min = df[column].min()
                column_max = df[column].max()
                if str(column_type).startswith('int'):
                    if column_min > np.iinfo(np.int8).min and column_max < np.iinfo(np.int8).max:
                        df[column] = df[column].astype(np.int8)
                    elif column_min > np.iinfo(np.int16).min and column_max < np.iinfo(np.int16).max:
                        df[column] = df[column].astype(np.int16)
                    elif column_min > np.iinfo(np.int32).min and column_max < np.iinfo(np.int32).max:
                        df[column] = df[column].astype(np.int32)
                    elif column_min > np.iinfo(np.int64).min and column_max < np.iinfo(np.int64).max:
                        df[column] = df[column].astype(np.int64)  
                else:
                    if column_min > np.finfo(np.float32).min and column_max < np.finfo(np.float32).max:
                        df[column] = df[column].astype(np.float32)
                    else:
                        df[column] = df[column].astype(np.float64)    
        # Размер занимаемой памяти после оптимизации (в Мб)
        after_size = df.memory_usage().sum() / 1024**2
        if print_size: print('Размер использования памяти: до {:5.2f} Mb - после {:5.2f} Mb ({:.1f}%)'
                             .format(before_size, after_size, 100 * (before_size - after_size) / before_size))
        return df
                                                                                                   
                                                                                                   
Что делает эта функция? Проверяет минимальное и максимальное значения в столбце. Если значения могут уместиться в типе данных меньшего размера (например, np.int8 вместо np.int16), столбец конвертируется в более компактный тип данных. Еще ты можешь обратить внимание на использование в функции [docstring](https://dzen.ru/a/Yw02qy-7BGdhFhN3?utm_referer=yandex.ru).  С её помощью мы оставляем описание того что делает эта функция, и эту подсказку мы можем вызвать в любом месте нашего проекта с помощью
    
    
    print(my_function.__doc__)
    
или    
    
    help(my_function)
 

    
Очень удобно когда функция сложна для восприятия, и нам нужно быстро получить подсказку что вообще эта функция делает и как    


In [9]:
# Оптимизация числовых типов данных
def optimize_memory_usage(df: pd.DataFrame, print_size: bool=True) -> pd.DataFrame:


    
    #Function optimizes memory usage in dataframe
    #df: pd.DataFrame - data table
    #print_size: bool - display of optimization results
    #return pd.DataFrame - amount of optimized memory
  


    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64'] # Типы, которые будем проверять на оптимизацию
    # Размер занимаемой памяти до оптимизации (в Мб)
    before_size = df.memory_usage().sum() / 1024**2    
    for column in df.columns:
        column_type = df[column].dtypes
        if column_type in numerics:
            column_min = df[column].min()
            column_max = df[column].max()
            if str(column_type).startswith('int'):
                if column_min > np.iinfo(np.int8).min and column_max < np.iinfo(np.int8).max:
                    df[column] = df[column].astype(np.int8)
                elif column_min > np.iinfo(np.int16).min and column_max < np.iinfo(np.int16).max:
                    df[column] = df[column].astype(np.int16)
                elif column_min > np.iinfo(np.int32).min and column_max < np.iinfo(np.int32).max:
                    df[column] = df[column].astype(np.int32)
                elif column_min > np.iinfo(np.int64).min and column_max < np.iinfo(np.int64).max:
                    df[column] = df[column].astype(np.int64)  
            else:
                if column_min > np.finfo(np.float32).min and column_max < np.finfo(np.float32).max:
                    df[column] = df[column].astype(np.float32)
                else:
                    df[column] = df[column].astype(np.float64)    
    # Размер занимаемой памяти после оптимизации (в Мб)
    after_size = df.memory_usage().sum() / 1024**2
    if print_size: print('Размер использования памяти: до {:5.2f} Mb - после {:5.2f} Mb ({:.1f}%)'
                         .format(before_size, after_size, 100 * (before_size - after_size) / before_size))
    return df
df= optimize_memory_usage(df)
df.info()

Размер использования памяти: до  1.53 Mb - после  1.08 Mb (29.2%)
<class 'pandas.core.frame.DataFrame'>
Index: 16669 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16669 non-null  object 
 1   platform         16669 non-null  object 
 2   year_of_release  16669 non-null  float32
 3   genre            16669 non-null  object 
 4   na_sales         16669 non-null  float32
 5   eu_sales         16669 non-null  float32
 6   jp_sales         16669 non-null  float32
 7   other_sales      16669 non-null  float32
 8   critic_score     16669 non-null  float32
 9   user_score       16669 non-null  float32
 10  rating           16669 non-null  object 
dtypes: float32(7), object(4)
memory usage: 1.1+ MB


### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>



Совет 🤔:


Опять же действия ниже правильные, но стоило предварительно вывести уникальные значения в столбце  жанры,  как бы мы увидели дубликаты. А ты сделал это только после того как скрытые дубликаты устранил  

In [10]:
# Проверяем уникальные значения в проблемных столбцах
unique_values = df['genre'].unique()
print(f"Уникальные значения в столбце genre: {unique_values}")

Уникальные значения в столбце genre: ['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']


In [11]:
#Приводим столбцы к нижнему/верхнему регистру
df['name'] = df['name'].str.lower()
df['genre'] = df['genre'].str.lower()
df['rating'] = df['rating'].str.upper()

In [12]:
#Подсчет уникальных названий игр
print('Кол-во уникальных игр:',df['name'].nunique())
#Поиск уникальных значений по столбцам
print('Уникальные платформы:',df['platform'].unique())
print('Уникальные жанры:',df['genre'].unique())
print('Уникальные года выпуска:',df['year_of_release'].unique())
print('Уникальные рейтинги:',df['rating'].unique())

Кол-во уникальных игр: 11422
Уникальные платформы: ['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']
Уникальные жанры: ['sports' 'platform' 'racing' 'role-playing' 'puzzle' 'misc' 'shooter'
 'simulation' 'action' 'fighting' 'adventure' 'strategy']
Уникальные года выпуска: [2006. 1985. 2008. 2009. 1996. 1989. 1984. 2005. 1999. 2007. 2010. 2013.
 2004. 1990. 1988. 2002. 2001. 2011. 1998. 2015. 2012. 2014. 1992. 1997.
 1993. 1994. 1982. 2016. 2003. 1986. 2000. 1995. 1991. 1981. 1987. 1980.
 1983.]
Уникальные рейтинги: ['E' 'NR' 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



👍  Всё как просят в проекте


 

<div class="alert alert-warning">
 


Совет 🤔:
 

Что можно было сделать еще?

Поискать неявные дубликаты   в рейтингах и названиях игр. Данные стали бы чуть чуть чище. 


На примере рейтингов это могло выглядеть так: 

1. Используем value_counts() к rating и обнаруживаем небольшое число у K-A и EC 
2. Гуглим и обнаруживаем что K-A это старая версия E (Кстати об этом у тебя было написано выше).. 
3. Делаем замены. Такой вариант работы с неявными дубликатами. 

 
    
Еще можно по отдельным столбцам:  .duplicated(subset=[а тут можно подумать какие столбцы использовать])    
    
    


 

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [13]:
#Создаем переменную с изначалальным датафреймом
first_df = len(df)
print(f'Кол-во строк до удаления дубликатов:{first_df}')
#Подсчет явных дубликатов
duplicates = df.duplicated()
print(f'Кол-во удаленных дубликатов:{duplicates.sum()}')
#Удаление явных дубликатов
df = df.drop_duplicates()
print(f'Кол-во строк после удаления дубликатов:{len(df)}')
#Доля удаленных дубликатов от изначального датафрейма
share = round(duplicates.sum()/first_df,2)*100
print(f'Доля удаленных строк:{share}')

Кол-во строк до удаления дубликатов:16669
Кол-во удаленных дубликатов:235
Кол-во строк после удаления дубликатов:16434
Доля удаленных строк:1.0


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



👍


Отдельный плюс за использование f-strings 
    
    
 
<div class="alert alert-warning">



Совет: 

- рекомендую вывести вокруг цифры какой то обьясняющий текст, лучше с помощью [f-stings](https://python-scripts.com/f-strings) , тут [свеженькое](https://habr.com/ru/company/wunderfund/blog/674866/) 
        

Было найдено 235 явных дубликатов и они были удалены из датафрейма. Доля удаленных строк от всех данных составляет 1 %

- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

- После проведения предобработки данных напишите общий промежуточный вывод.

В датафрейме у нескольких столбцов был тип данных, который не позволял проводит какие-либо расчеты, были обнаружены несущественные значения для анализа, так же были найдены дубликаты, которые нужно было удалить из нашего датафрейма. Все вышепоставленные задачи были выполнены.

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [14]:
#Отбор данных за период с 2000 по 2013 год
df_actual = df[(df['year_of_release']>=2000) & (df['year_of_release']<=2013)].copy() 
#Проверяем размер и первые несколько строк нового датафрейма
display(df_actual.shape)
display(df_actual.head())

(12772, 11)

name platform  year_of_release     genre   na_sales  \
0             wii sports      Wii           2006.0    sports  41.360001   
2         mario kart wii      Wii           2008.0    racing  15.680000   
3      wii sports resort      Wii           2009.0    sports  15.610000   
6  new super mario bros.       DS           2006.0  platform  11.280000   
7               wii play      Wii           2006.0      misc  13.960000   

    eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0  28.959999      3.77         8.45          76.0         8.0      E  
2  12.760000      3.79         3.29          82.0         8.3      E  
3  10.930000      3.28         2.95          80.0         8.0      E  
6   9.140000      6.50         2.88          89.0         8.5      E  
7   9.180000      2.93         2.84          58.0         6.6      E

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>



Совет 🤔:


    df_actual = df[(df['year_of_release']>=2000) & (df['year_of_release']<=2013)]  -----> df_actual = df[(df['year_of_release']>=2000) & (df['year_of_release']<=2013)].copy() 
    
    
И тогда мы сможем "развести"  df_actual  и df, а пока что  любые изменения которые ты сделаешь в df_actual,  приведут к изменениям в   df, а нам это не всегда нужно (Это происходит потому что без .copy  настоящая копия не создаётся, и df_actual    df ссылаются на одно место в памяти,  поэтому по-прежнему будут меняться одновременно) 

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [15]:
#Задаем границы интервалов
bins = [-5,0,3,8,10]
#Задаем категории оценивания
labels = ['отсутсвует оценка','низкая оценка', 'средняя оценка', 'высокая оценка']
#Создаем новый столбец 'user_rating'
df['user_rating'] = pd.cut(df['user_score'], bins=bins,labels=labels,right=False,include_lowest=True)
#Группируем по оценкам пользователей и ведем подсчет игр
group = df.groupby('user_rating',observed=False).agg({'name':'count', 'user_score':'mean'}).reset_index()
group.columns = ['user_rating','amount_of_games','avg_count']
display(group)

user_rating  amount_of_games  avg_count
0  отсутсвует оценка             8979  -1.000000
1      низкая оценка              142   2.080282
2     средняя оценка             4773   6.550995
3     высокая оценка             2540   8.486496

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



👍 right=False

Молодец
 
    
    
<div class="alert alert-warning">



Совет 🤔:


альтернатива использовать [pivot_table](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html) - продвинутый инструмент, позволяющий делать в том числе агрегацию (по умолчанию считает среднюю). Очень хороший инструмент, можешь попробовать его, в будущем он явно пригодится (по сути это полная аналогия сводных таблиц в excell)

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [16]:
#Задаем границы интервалов
bins = [-5,0,30,80,100]
#Задаем категории оценивания
labels = ['отсутсвует оценка','низкая оценка', 'средняя оценка', 'высокая оценка']
#Создаем новый столбец 'critic_rating'
df['critic_rating'] = pd.cut(df['critic_score'], bins=bins,labels=labels,right=False)
#Группируем по оценкам критиков и ведем подсчет игр
group = df.groupby('critic_rating',observed=False).agg({'name':'count','critic_score':'mean'}).reset_index()
group.columns = ['critic_rating','amount_of_games','avg_count']
display(group)

critic_rating  amount_of_games  avg_count
0  отсутсвует оценка             8459  -1.000000
1      низкая оценка               59  24.694916
2     средняя оценка             5944  64.056190
3     высокая оценка             1972  85.151619

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [17]:
#Группируем по платформе и ведем подсчет игр на этой платформе
group_platform = df_actual.groupby('platform')['name'].count()
#Сортируем результат
top_7_plat = group_platform.sort_values(ascending = False).head(7)
display(top_7_plat)

platform
PS2     2126
DS      2117
Wii     1275
PSP     1179
X360    1118
PS3     1087
GBA      810
Name: name, dtype: int64

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



👍Наглядно


 

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

В результате выполнения всех поставлених задач, в датасет у нас добавилось 2 столбца:**user_rating и critic_rating** тип данных у обоих является 'category'. В этих значениях мы получили данные о количестве игр с разными категориями оценок. Как мы можем видеть, очень мало пользователей ставят маленьку оценку играм, всего 142 игры относится к категории *'низкая оценка'*. Судя по имеющимся у нас данным, людям **нравится играть в игры**.Теперь рассмотрим столбец **critic_rating**,где оценки играм проставляли уже критики, всего 59 игр подходит под категорию *'низкая оценка'*,это в **2.5** раза меньше, чем у обычных пользователей. Какой итог по оценкам игр со стороны пользователей и критиков мы можем подвести? Разработка игр не стоит на месте, сфера развивается и с такой положительной статистикой хочется, чтоб разработчики продолжали радовать всех интересными и качественным играми.

Судя по срезу, где у нас предоставлена информация о играх на разных платфоормах, лидирует **PS2**, *2126* игра может быть установлена на данной платформе, это достаточно много, его предшественик,**DS**, ушел недалеко, ведь он может скачивать *2217* игр. Эта статистика говорит, что данные платформы на момент выпуска и ее использования были в пике своей продаваемости, и среда по разработке игр не стояла на месте, поэтому на эти 2 платформы находятся сейчас дальше всех и являются **самыми востребоваными в мире**.

 



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Александр, здорово что в конце расписаны   этапы работы и полученные результаты. Это важно потому что когда проект захочет посмотреть будущий работодатель (или начальник), у него может не быть времени на подробный разбор кода. Вероятнее всего он бегло просмотрит код, а из общего вывода захочет получить представление о всей работе.

 


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>


Александр, у тебя старательно выполненная работа, все четко, осмысленно. Выводы присутствуют, они четкие и подробные. В своих действиях используешь продвинутые оптимальные анализа,  подробно комментируешь свои действия. При работе с пропусками принял верные решение

 
    
Я оставил небольшие советы, если есть желание и время можешь воспользоваться    
    

    
    
Что стоит исправить:

 
    
- при знакомстве   стоит вывести на экран наши данные, это может подсказать что нам стоит с ними предпринять в дальнейшем 
    
    
- перед исправлением типа данных стоит вывести уникальные значения по столбцам где мы видим несоответствия,  и проанализировать увиденное    
    
    
    
    
Жду исправлений, для принятия проекта. Если какие то вопросы, то сразу спрашивай ) 


</div>


***Спасибо за комментарии***

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:


Я бы посоветовал оставлять комментарии, когда делаешь исправление.  Так коммуникация будет лучше  


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Спасибо за работу!    
 
Александр, тобой исправлено красное, желтое не тронуто - твое право ) Тем не менее надеюсь мои советы и вопросики были полезны и ты узнала что то новое.
Отличная работа. Желаю успехов в дальнейшей учебе!
  